domain/network/state/subnet.go

```diff
@@ -0,0 +1,440 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	"github.com/juju/juju/core/network"
+)
+
+const (
+	subnetTypeBase              = 0
+	subnetTypeFanOverlaySegment = 1
+)
+
+// AddSubnet creates a subnet.
+func (st *State) AddSubnet(
```

Comment:
Why don't we accept `SubnetInfo` like the current `state` method?

Reply:
We can discuss this more globally, but IMO it's more clear and explicit to have each db field (to be inserted/updated) as input argument (when possible, if not too many) on the state layer methods.
For example if we accepted `SpaceInfo` here, one might think that each of the subnets in the struct will be updated with the values contained in that slice, while it won't.